resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkheadConfig.java

```diff
@@ -0,0 +1,127 @@
+/*
+ *
+ *  Copyright 2019: Bohdan Storozhuk, Mahmoud Romeh
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.bulkhead.adaptive;
+
+import java.util.function.Predicate;
+
+import io.github.resilience4j.bulkhead.adaptive.internal.AdaptiveLimitBulkhead;
+import io.github.resilience4j.bulkhead.adaptive.internal.config.AIMDConfig;
+import io.github.resilience4j.core.lang.Nullable;
+
+/**
+ * A {@link AdaptiveBulkheadConfig} configures a adaptation capabilities of  {@link AdaptiveLimitBulkhead}
+ */
+public class AdaptiveBulkheadConfig<T> {
+	private T config;
+	@Nullable
+	private Predicate<Exception> adaptIfError;
+	private int initialConcurrency = 1;
+
+	private AdaptiveBulkheadConfig() {
+	}
+
+	public int getInitialConcurrency() {
+		return initialConcurrency;
+	}
+
+	@Nullable
+	public T getConfiguration() {
+		return config;
+	}
+
+	@Nullable
+	public Predicate<Exception> getAdaptIfError() {
+		return adaptIfError;
+	}
+
+	/**
+	 * Returns a builder to create a custom AdaptiveBulkheadConfig.
+	 *
+	 * @return a {@link AdaptiveBulkheadConfig.Builder}
+	 */
+	public static <T> Builder<T> from(AdaptiveBulkheadConfig<T> baseConfig) {
+		return AdaptiveBulkheadConfig.builder(baseConfig);
+	}
+
+	/**
+	 * Creates a default Bulkhead configuration.
+	 *
+	 * @return a default Bulkhead configuration.
+	 */
+	public static AdaptiveBulkheadConfig<AIMDConfig> ofDefaults() {
+		return AdaptiveBulkheadConfig.<AIMDConfig>builder().config(AIMDConfig.builder().build()).build();
+	}
+
+	/**
+	 * Returns a builder to create a custom AdaptiveBulkheadConfig.
+	 *
+	 * @return a {@link AdaptiveBulkheadConfig.Builder}
+	 */
+	public static <T> Builder<T> builder() {
+		return new Builder<>();
+	}
+
+	/**
+	 * Returns a builder to create a custom AdaptiveBulkheadConfig.
+	 *
+	 * @return a {@link AdaptiveBulkheadConfig.Builder}
+	 */
+	public static <T> Builder<T> builder(AdaptiveBulkheadConfig<T> bulkheadConfig) {
+		return new Builder<>(bulkheadConfig);
+	}
+
+	public static class Builder<T> {
+		private final AdaptiveBulkheadConfig<T> adaptiveBulkheadConfig;
+
+		private Builder() {
+			adaptiveBulkheadConfig = new AdaptiveBulkheadConfig<>();
+		}
+
+		private Builder(AdaptiveBulkheadConfig<T> adaptiveBulkheadConfig) {
+			this.adaptiveBulkheadConfig = adaptiveBulkheadConfig;
+		}
+
+		public Builder<T> adaptIfError(Predicate<Exception> adaptIfError) {
```

Comment:
I think we should implement it similar to the CircuitBreaker.
Some exceptions can be totally ignored. Some exceptions count as a success and some exceptions should increase the failure rate.

Reply:
Means: We need `ignoreExceptions` and `recordExceptions`.